content/post/asdf.md

```diff
@@ -0,0 +1,240 @@
+models from the users. That is, it solves the following two major problems
+every ML engineer hits:
+
+1. What to save a trained model and load it back.
+2. How to reference and distribute trained models.
+
+Let's not go offtopic with discussing (2) - there are many funny subproblems there
+which deserve a dedicated blog post about Modelforge. Instead, the focus will be on
+serialization. The Python crowd is used to solve (1) with
+[`pickle`](https://docs.python.org/3/library/pickle.html) - the built-in
+serialization module. It works nicely while these three conditions hold:
+
+1. The origin of the files is trusted and secure.
+2. The interoperability with other languages is not needed.
+3. The serialized objects are not big.
+
+As some know very well, `pickle` contains a full-featured virtual machine inside,
+which is interpreted to recreate the serialized Python objects. The advantage
+is that we can serialize literally any object, except for few rare types
+like mutexes or executable code. The implied drawbacks are substantial though:
+we should not load a pickle from unreliable source because it is vulnerable
+to remote code execution attacks; the loader must implement a virtual machine,
+which is hard with other programming languages; the memory consumption
+during serialization and deserialization can grow far beyond the original object size.
+The latter is important for machine learning and data science in particular:
+the objects can grow very big, say, bigger than a few hundred megabytes and
+pickling them is very slow and very demanding for the free operating memory.
+We witnessed 2 and even 3 times bigger memory consumption during
+`pickle`-ing which led to funny situtations when one has successfully computed
+the data but cannot dump it on disk so the whole effort becomes useless.
+
+So when we started Modelforge, we searched for the best serialization format
+which is **not** `pickle`.
+
+## Life beyond `pickle`
+
+Our best-shot requirements were:
+
+1. Binary format. It is impossible to efficiently save a huge dense tensor in JSON or YAML.
+2. At the same time, it would be nice to be able to save any JSON-like metadata without pain.
+3. The schema is optional. We did not want to force people to write the schema definitions - it is asking too much from ML researchers. Schema-ful formats separate it from the data and don't include
+into the file, and thus it becomes hard to introspect. Pythonistas take introspection for granted, to say at least.
+4. The typical data size may span over tens of gigabytes. Compared to gigabytes,
+the overhead of including a schema looks irrelevant indeed, hence the files
+become self-descriptive and the schema turns into a validation perk.
+This requirement rules out Protocol Buffers completely, because there can be
+performance and memory problems after 50MB already. It also excludes other
+popular binary serialization formats because they always target messaging and RPC.
```

Comment:
protocol buffers are initially designed for storage, not messaging
gRPC is built on top for messaging

Reply:
That's very interesting, I would love to read about the origins of PB - I have always thought that messaging was the main goal.